src/pe/tls.rs

```diff
@@ -0,0 +1,247 @@
+                    itd.start_address_of_raw_data,
+                    itd.end_address_of_raw_data
+                )));
+            }
+
+            if (itd.start_address_of_raw_data as usize) < image_base {
+                return Err(error::Error::Malformed(format!(
+                    "tls start_address_of_raw_data ({:#x}) is less than image base ({:#x})",
+                    itd.start_address_of_raw_data, image_base
+                )));
+            }
+
+            // VA to RVA
+            let rva = itd.start_address_of_raw_data as usize - image_base;
+            let size = itd.end_address_of_raw_data - itd.start_address_of_raw_data;
+            let offset =
+                utils::find_offset(rva, sections, file_alignment, opts).ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "cannot map tls start_address_of_raw_data rva ({:#x}) into offset",
+                        rva
+                    ))
+                })?;
+            raw_data = Some(&bytes[offset..offset + size as usize]);
+        }
+
+        // Parse the index if any
+        if itd.address_of_index != 0 {
+            if (itd.address_of_index as usize) < image_base {
+                return Err(error::Error::Malformed(format!(
+                    "tls address_of_index ({:#x}) is less than image base ({:#x})",
+                    itd.address_of_index, image_base
+                )));
+            }
+
+            // VA to RVA
+            let rva = itd.address_of_index as usize - image_base;
+            let offset =
+                utils::find_offset(rva, sections, file_alignment, opts).ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "cannot map tls address_of_index rva ({:#x}) into offset",
+                        rva
+                    ))
+                })?;
+
+            slot = Some(bytes.pread_with::<u32>(offset, scroll::LE)?);
+        }
+
+        // Parse the callbacks if any
+        if itd.address_of_callbacks != 0 {
+            if (itd.address_of_callbacks as usize) < image_base {
+                return Err(error::Error::Malformed(format!(
+                    "tls address_of_callbacks ({:#x}) is less than image base ({:#x})",
+                    itd.address_of_callbacks, image_base
+                )));
+            }
+
+            // VA to RVA
+            let rva = itd.address_of_callbacks as usize - image_base;
+            let offset =
+                utils::find_offset(rva, sections, file_alignment, opts).ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "cannot map tls address_of_callbacks rva ({:#x}) into offset",
+                        rva
+                    ))
+                })?;
+            let mut i = 0;
+            // Read the callbacks until we find a null terminator
+            loop {
```

Comment:
on a malformed binary could this cause an infinite loop?

Reply:
Yes. I think it would make sense to have a RVA validation there and return the malformed error if invalid RVAs found.

Reply:
Done.

Reply:
so what i meant specifically is: is this loop guaranteed to terminate? this comes up when there are malformed binaries, and the loop only terminates in a well-formed case, but if the loop has no upper bound, it might cause the library to run forever, which is unacceptable.

Reply:
Yea I was just making sure there wasn't some obvious termination condition we could add, e.g., end of file, etc., that would guarantee we are done looping; my general reflex when i see `loop` is to ask how it can terminate, as we've had bugs in past of infinite loops without proper termination condition, but your answer is sufficient, thank you!